### PR TITLE
CEED_STRIDES_BACKEND optimization for cuda-ref operator apply

### PR DESCRIPTION
**Pull request by nbeams**
The PR is for discussion of an optimization for the cuda-ref and magma backends, avoiding unnecessary calls to strided restriction when using `CEED_STRIDES_BACKEND`. 

The goal was to have something that works for both cuda-ref and magma with as little code duplication as possible, despite the fact that the two backends use different E-Vector layouts.

The commit 8d5939ec2b201c656fdfeb719b7b35d54e197953 contains the version I feel more confident is acceptable.  It adds a strided restriction for the NONE eval mode in the output basis action, which creates a properly-ordered E-Vector.  The optimization happens on the input side, where inputs are checked during operator setup for being: passive, with `NONE` basis emode, a strided restriction, and `CEED_STRIDES_BACKEND`.  If all of these are true, the E-Vector for this field is never created, and during the input element restriction, the input vector for the field is read directly to the `edata`.  The end result of this for the usual tests or BPs is two strided restriction calls--one each in `TRANSPOSE` and `NO_TRANPOSE` mode--during output of the build operator, and no calls for the input of the Apply operator. 

Commit 37474e8143852713feafc80abfc481ee0675368f was an attempt to add one more step of optimization for the magma backend, which uses the same layout for E- and Q-Vectors (component as slowest index, then elements, then nodes) and does not need the extra element restriction call during the `NONE` output basis action, but can set the pointer directly.  In order to do this, I added a new boolean variable for the `CeedOperator_Cuda` struct.  It gets set to `true` in `CeedOperatorCreate_Cuda`.  Then I added a `CeedOperatorCreate_Magma` function to the magma backend, which actually just calls the cuda-ref version, and then explicitly sets this boolean variable to false.  This variable is then used to choose different output options in the operator application.  The end result is just one strided restriction (`TRANSPOSE` mode) during the setup operator, instead of two, for the magma backend.

However, I'm not sure if this is an acceptable "crossing" of backends, or if it creates problems I haven't thought of yet.

### CODE REVIEW

**Comment by jeremylt** [comment]
Travis is failing because of `make style` changes. I'm going to try `make tidy` with the fixes from my other branch.

File: backends/cuda/ceed-cuda-operator.c [hunk 1]
```diff
@@ -78,12 +82,41 @@ static int CeedOperatorSetupFields_Cuda(CeedQFunction qf, CeedOperator op,
...
+
+      // First, check whether the field is input or output:
+      if (!inOrOut) {
+        // Check for passive input:
+        ierr = CeedOperatorFieldGetVector(opfields[i], &field_vec); CeedChk(ierr);
+        if (field_vec != CEED_VECTOR_ACTIVE) {
+          // Check emode
+          if (emode == CEED_EVAL_NONE) {
+            // Check for strided restriction
+            ierr = CeedElemRestrictionGetStridedStatus(Erestrict, &is_strided);
```
Comment (jeremylt) [inline backends/cuda/ceed-cuda-operator.c:103]: Need `CeedChk(ierr);`

File: backends/cuda/ceed-cuda-operator.c [hunk 2]
```diff
@@ -78,12 +82,41 @@ static int CeedOperatorSetupFields_Cuda(CeedQFunction qf, CeedOperator op,
...
+            if (is_strided) {
+              // Check if vector is already in preferred backend ordering
+              ierr = CeedElemRestrictionGetBackendStridesStatus(Erestrict, &skip_restrict);
```
Comment (jeremylt) [inline backends/cuda/ceed-cuda-operator.c:106]: `CeedChk(ierr);`

File: backends/cuda/ceed-cuda-operator.c [hunk 3]
```diff
@@ -212,17 +245,25 @@ static int CeedOperatorApplyAdd_Cuda(CeedOperator op, CeedVector invec,
...
-        vec = invec;
-      // Restrict
+      // Get input element restriction
       ierr = CeedOperatorFieldGetElemRestriction(opinputfields[i], &Erestrict);
       CeedChk(ierr);
-      ierr = CeedElemRestrictionApply(Erestrict, CEED_NOTRANSPOSE, vec,
+      if (vec == CEED_VECTOR_ACTIVE) 
+        vec = invec;
+      // Restrict, if necessary
+      if (impl->evecs[i] == NULL) {
```
Comment (jeremylt) [inline backends/cuda/ceed-cuda-operator.c:254]: or `!impl->evecs[i]` if you prefer?

File: backends/cuda/ceed-cuda.h [hunk 4]
```diff
@@ -113,6 +113,7 @@ typedef struct {
   CeedVector *qvecsout;   /// Output Q-vectors needed to apply operator
   CeedInt    numein;
   CeedInt    numeout;
+  bool e_and_q_layout_diff;
```
Comment (jeremylt) [inline backends/cuda/ceed-cuda.h:116]: The style in the CUDA backends would be to use `eandqlayoutdiff`  or maybe `eandqdiffer`?

File: backends/magma/ceed-magma-operator.c [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+// testbed platforms, in support of the nation's exascale computing imperative.
+
+#include <ceed-backend.h>
+#include "../cuda/ceed-cuda.h"
+#include "ceed-magma.h"
+#include <string.h>
+
+// Create a cuda-ref operator and modify its e_and_q_layout_diff value
+int CeedOperatorCreate_Magma(CeedOperator op) {
+
```
Comment (jeremylt) [inline backends/magma/ceed-magma-operator.c:24]: extra line

File: backends/magma/ceed-magma-operator.c [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+//
+// The CEED research is supported by the Exascale Computing Project 17-SC-20-SC,
+// a collaborative effort of two U.S. Department of Energy organizations (Office
+// of Science and the National Nuclear Security Administration) responsible for
+// the planning and preparation of a capable exascale ecosystem, including
+// software, applications, hardware, advanced system engineering and early
+// testbed platforms, in support of the nation's exascale computing imperative.
+
+#include <ceed-backend.h>
+#include "../cuda/ceed-cuda.h"
```
Comment (jeremylt) [inline backends/magma/ceed-magma-operator.c:18]: Some day we'll add a way to make this sort of thing less ugly looking. I like this approach though.

**Comment by jeremylt** [comment]
I'm getting nek and mfem example failures for BP3 with only `/gpu/magma` on my local machine. They aren't happening on master. `/cpu/cuda` is fine though.

To replicate
```
make prove BACKENDS="/gpu/magma" -j4 search=mfem PROVE_OPTS=-v
make prove BACKENDS="/gpu/magma" -j4 search=nek PROVE_OPTS=-v
```

File: backends/cuda/ceed-cuda-operator.c [hunk 7]
```diff
@@ -61,6 +61,10 @@ static int CeedOperatorSetupFields_Cuda(CeedQFunction qf, CeedOperator op,
   CeedElemRestriction Erestrict;
   CeedOperatorField *opfields;
   CeedQFunctionField *qffields;
+  CeedVector field_vec;
+  bool is_strided;
+  bool skip_restrict;
```
Comment (jeremylt) [inline backends/cuda/ceed-cuda-operator.c:66]: Similar naming convention comment. Personally, I'm a fan of `camelCase` and its in the CPU backends. I don't think we use `snake_case` anywhere but the Python code, except for using `d_foo` and `h_foo`.

**Comment by nbeams** [comment]
> I'm getting nek and mfem example failures for BP3 with only `/gpu/magma` on my local machine. They aren't happening on master. `/cpu/cuda` is fine though.
> 
> To replicate
> 
> ```
> make prove BACKENDS="/gpu/magma" -j4 search=mfem PROVE_OPTS=-v
> make prove BACKENDS="/gpu/magma" -j4 search=nek PROVE_OPTS=-v
> ```

Hmm, this is concerning me.  I tried running the bp3 benchmark I already had set up with the updated fork of MFEM and updated branch of libCEED, and it was fine.  When I try to build the mfem examples directly, I'm getting build errors: 
`nvcc fatal   : Unknown option 'Wl,-rpath,/home/nbeams/CEED/libCEED/lib'`

I feel like there's probably something really simple I need to configure or change...?

**Comment by jedbrown** [comment]
@nbeams Can you add `V=1` to your make and see what command is used? I'm not sure why `nvcc` is seeing a linker flag since I wouldn't expect `nvcc` to be used as the linker.

File: backends/cuda/ceed-cuda.h (hunk 4 above)
Comment (YohannDudouit) [reply backends/cuda/ceed-cuda.h:116]: I'm not a very consistent style person (unfortunately), so you can unify the way you prefer.

**Comment by nbeams** [comment]
It seems that `MFEM_CXX` is `nvcc` in my `config.mk` for MFEM and that's what's used in the command in examples/mfem/Makefile. 

make[1]: Entering directory `/home/nbeams/CEED/bmrkNT_icl_opt/benchmarks/builds/linux_gcc/libceed/examples/mfem'
/spack/opt/spack/linux-scientific7-x86_64/gcc-7.3.0/cuda-9.2.88-idc7ndod3qxkas6ybndjp6wmi4pguvnh/bin/nvcc -I/home/nbeams/CEED/bmrkNT_icl_opt/benchmarks/builds/linux_gcc/libceed/include -O -g  -std=c++11 -Xcompiler="-O3 -mcmodel=medium" -x=cu --expt-extended-lambda -arch=sm_60 -ccbin mpicxx -I/home/nbeams/CEED/bmrkNT_icl_opt/benchmarks/builds/linux_gcc/mfem -I/home/nbeams/CEED/bmrkNT_icl_opt/benchmarks/builds/linux_gcc/hypre/src/hypre/include  -I/home/nbeams/CEED/bmrkNT_icl_opt/benchmarks/builds/linux_gcc/libceed/include -DMFEM_DIR="\"/home/nbeams/CEED/bmrkNT_icl_opt/benchmarks/builds/linux_gcc/mfem\"" /home/nbeams/CEED/bmrkNT_icl_opt/benchmarks/builds/linux_gcc/libceed/examples/mfem/bp1.cpp -o bp1 \
  -Wl,-rpath,/home/nbeams/CEED/bmrkNT_icl_opt/benchmarks/builds/linux_gcc/libceed/lib -L/home/nbeams/CEED/bmrkNT_icl_opt/benchmarks/builds/linux_gcc/libceed/lib -lceed -L/home/nbeams/CEED/bmrkNT_icl_opt/benchmarks/builds/linux_gcc/mfem -lmfem -L/home/nbeams/CEED/bmrkNT_icl_opt/benchmarks/builds/linux_gcc/hypre/src/hypre/lib -lHYPRE  -L/home/nbeams/CEED/bmrkNT_icl_opt/benchmarks/builds/linux_gcc/metis-4.0.3 -lmetis -Xlinker=-rpath,/home/nbeams/CEED/bmrkNT_icl_opt/benchmarks/builds/linux_gcc/libceed/lib -L/home/nbeams/CEED/bmrkNT_icl_opt/benchmarks/builds/linux_gcc/libceed/lib -lceed  -lrt

This is the libCEED/MFEM pair I have installed in a directory for testing the benchmarks (though both have been updated) -- maybe I should try building a stand-alone MFEM.

**Comment by v-dobrev** [comment]
@nbeams, when using `nvcc` for linking, parameters that start with `-Wl,` have to be replaced to start with `-Xlinker=`. One way to make this work in the `Makefile`, is to replace the line https://github.com/CEED/libCEED/blob/aa2c314791b491352859d68bb43216cc2a7b5249/examples/mfem/Makefile#L22
with
```makefile
CEED_LIBS ?= $(XLINKER)-rpath,$(abspath $(CEED_DIR)/lib) -L$(CEED_DIR)/lib -lceed
```
and then after the line https://github.com/CEED/libCEED/blob/aa2c314791b491352859d68bb43216cc2a7b5249/examples/mfem/Makefile#L31

add something like this:
```makefile
ifeq (YES,$(MFEM_USE_CUDA))
   XLINKER = -Xlinker=
else
   XLINKER = -Wl,
endif
```

**Comment by nbeams** [comment]
Thanks, @v-dobrev, that did fix the linking issue.  But then I get errors like 

`In function ‘int f_build_mass(void*, CeedInt, const CeedScalar* const*, CeedScalar* const*)’:`
`examples/mfem/bp1.h:33:1: error: for, while or do statement expected before ‘case’`
`   case 11:`
` ^ ~~`

**Comment by nbeams** [comment]
@jeremylt, can you provide any more info about the failures of the MFEM tests (I know make prove doesn't show the output)?  I started with a clean build of your MFEM fork.  I have run `mfem-bp3` for /cpu/self/opt, /gpu/cuda/ref, and /gpu/magma (default options for other parameters) and they all have the same output (333 iterations, reduction factor 0.959126).  The operator setup in the *.hpp files doesn't use `CEED_STRIDES_BACKEND`, so the optimization isn't seen when I look at nvprof output for the two GPU backends.

I also tried running `ex1` from MFEM's examples, and it gives the same output for ceed-cpu and cuda-ref/magma backends.  There, I do see the reduction in the extra strided restriction calls when profiling, because the setup in `ceed.cpp` uses `CEED_STRIDES_BACKEND`.

I don't have Nek built on this machine, so I haven't tested that yet.

**Comment by jeremylt** [comment]
To see output from `make prove` you can run it with `make prove PROVE_OPTS=-v`.

Are you running the MFEM bp3 with the test arguments?
`-ceed {ceed_resource} -t -no-vis --size 2000`?

**Comment by nbeams** [comment]
No, I guess the default options that are different than the test options are:
```
   --size 50000
   --visualization
   --no-test
```

I just ran your make prove command above and the output was:
```
make: 'lib' with optional backends: /gpu/magma
Testing backends: /gpu/magma
prove -v --exec 'tests/tap.sh' mfem-bp1 mfem-bp3
mfem-bp1 .. 
1..3
# TESTARGS: -ceed {ceed_resource} -t -no-vis --size 2000
ok 1 mfem-bp1 /gpu/magma
ok 2 mfem-bp1 /gpu/magma stdout
ok 3 mfem-bp1 /gpu/magma stderr
ok
mfem-bp3 .. 
1..3
# TESTARGS: -ceed {ceed_resource} -t -no-vis --size 2000
ok 1 mfem-bp3 /gpu/magma
ok 2 mfem-bp3 /gpu/magma stdout
ok 3 mfem-bp3 /gpu/magma stderr
ok
```

**Comment by jeremylt** [comment]
Your last commit now works for me
```
[jeremy@sinensis libCEED]$ make prove search=mfem PROVE_OPTS=-v BACKENDS="/gpu/magma /gpu/cuda/ref"
make: 'lib' with optional backends: /gpu/magma /gpu/cuda/ref
make -C examples/mfem CEED_DIR=`pwd` \
  MFEM_DIR="/home/jeremy/Dev/mfem" bp1
make[1]: Entering directory '/home/jeremy/Dev/libCEED/examples/mfem'
g++ -I/home/jeremy/Dev/libCEED/include -O -g  -O3 -std=c++11 -I/home/jeremy/Dev/mfem  -DMFEM_DIR="\"/home/jeremy/Dev/mfem\"" /home/jeremy/Dev/libCEED/examples/mfem/bp1.cpp -o bp1 \
  -Wl,-rpath,/home/jeremy/Dev/libCEED/lib -L/home/jeremy/Dev/libCEED/lib -lceed -L/home/jeremy/Dev/mfem -lmfem -lrt
make[1]: Leaving directory '/home/jeremy/Dev/libCEED/examples/mfem'
mv examples/mfem/bp1 build/mfem-bp1
make -C examples/mfem CEED_DIR=`pwd` \
  MFEM_DIR="/home/jeremy/Dev/mfem" bp3
make[1]: Entering directory '/home/jeremy/Dev/libCEED/examples/mfem'
g++ -I/home/jeremy/Dev/libCEED/include -O -g  -O3 -std=c++11 -I/home/jeremy/Dev/mfem  -DMFEM_DIR="\"/home/jeremy/Dev/mfem\"" /home/jeremy/Dev/libCEED/examples/mfem/bp3.cpp -o bp3 \
  -Wl,-rpath,/home/jeremy/Dev/libCEED/lib -L/home/jeremy/Dev/libCEED/lib -lceed -L/home/jeremy/Dev/mfem -lmfem -lrt
make[1]: Leaving directory '/home/jeremy/Dev/libCEED/examples/mfem'
mv examples/mfem/bp3 build/mfem-bp3
Testing backends: /gpu/magma /gpu/cuda/ref
prove -v --exec 'tests/tap.sh' mfem-bp1 mfem-bp3
mfem-bp1 .. 
1..6
# TESTARGS: -ceed {ceed_resource} -t -no-vis --size 2000
ok 1 mfem-bp1 /gpu/magma
ok 2 mfem-bp1 /gpu/magma stdout
ok 3 mfem-bp1 /gpu/magma stderr
ok 4 mfem-bp1 /gpu/cuda/ref
ok 5 mfem-bp1 /gpu/cuda/ref stdout
ok 6 mfem-bp1 /gpu/cuda/ref stderr
ok
mfem-bp3 .. 
1..6
# TESTARGS: -ceed {ceed_resource} -t -no-vis --size 2000
ok 1 mfem-bp3 /gpu/magma
ok 2 mfem-bp3 /gpu/magma stdout
ok 3 mfem-bp3 /gpu/magma stderr
ok 4 mfem-bp3 /gpu/cuda/ref
ok 5 mfem-bp3 /gpu/cuda/ref stdout
ok 6 mfem-bp3 /gpu/cuda/ref stderr
ok
All tests successful.
Files=2, Tests=12,  5 wallclock secs ( 0.01 usr  0.00 sys +  2.80 cusr  2.14 csys =  4.95 CPU)
Result: PASS
```

**Comment by nbeams** [comment]
That's weird -- I haven't committed anything since I opened the PR? 
Oh well, glad it's passing now.  Are the Nek tests still failing?  I will have to work on getting that set up to test.

**Comment by jeremylt** [comment]
Looks good on my end
```
[jeremy@sinensis libCEED]$ make prove search=nek PROVE_OPTS=-v BACKENDS="/gpu/magma /gpu/cuda/ref"
make: 'lib' with optional backends: /gpu/magma /gpu/cuda/ref
Testing backends: /gpu/magma /gpu/cuda/ref
prove -v --exec 'tests/tap.sh' nek-bps
nek-bps .. 
1..12
# TESTARGS: -c {ceed_resource} -e bp1 -n 1 -b 4 -test
ok 1 nek-bps /gpu/magma
ok 2 nek-bps /gpu/magma stdout
ok 3 nek-bps /gpu/magma stderr
ok 4 nek-bps /gpu/cuda/ref
ok 5 nek-bps /gpu/cuda/ref stdout
ok 6 nek-bps /gpu/cuda/ref stderr
# TESTARGS: -c {ceed_resource} -e bp3 -n 1 -b 4 -test
ok 7 nek-bps /gpu/magma
ok 8 nek-bps /gpu/magma stdout
ok 9 nek-bps /gpu/magma stderr
ok 10 nek-bps /gpu/cuda/ref
ok 11 nek-bps /gpu/cuda/ref stdout
ok 12 nek-bps /gpu/cuda/ref stderr
ok
All tests successful.
Files=1, Tests=12,  7 wallclock secs ( 0.01 usr  0.00 sys +  3.40 cusr  2.26 csys =  5.67 CPU)
Result: PASS
```

File: backends/cuda/ceed-cuda.h (hunk 4 above)
Comment (nbeams) [reply backends/cuda/ceed-cuda.h:116]: I like `eandqdiffer`, I think I will go with that.

File: backends/cuda/ceed-cuda-operator.c (hunk 2 above)
Comment (nbeams) [reply backends/cuda/ceed-cuda-operator.c]: These were added in 8a0ea318b2c4645d3b032728c961f814ba2c3fdd

File: backends/cuda/ceed-cuda-operator.c (hunk 3 above)
Comment (nbeams) [reply backends/cuda/ceed-cuda-operator.c]: Yes, I agree that's cleaner, I changed it.

File: backends/magma/ceed-magma-operator.c (hunk 5 above)
Comment (nbeams) [reply backends/magma/ceed-magma-operator.c]: Fixed in 3026fe3dfa6c1d826d17c138bee5c49103da2abb

File: backends/cuda/ceed-cuda-operator.c (hunk 7 above)
Comment (nbeams) [reply backends/cuda/ceed-cuda-operator.c]: MAGMA itself uses `snake_case`, but you're right that I should match the CUDA backend style in this file.  I changed `is_strided` to `strided`, `skip_restrict` to `skiprestrict`, and `field_vec` to `fieldvec`, which I think "blends in" more.

**Comment by nbeams** [comment]
I have now built and run the Nek examples locally and they are passing for magma and cuda-ref (though they don't benefit from the optimization because they don't use `CEED_STRIDES_BACKEND`). MFEM's `ex1` does have the optimization, and works locally for me as well.  Also, I think I have addressed all of @jeremylt's review comments.

**Comment by jeremylt** [comment]
Woops, I am mistaken. Are you seeing this?
```
[jeremy@sinensis libCEED]$ make prove -j4 BACKENDS="/gpu/magma /gpu/cuda/ref"  search=t536 PROVE_OPTS=-v
make: 'lib' with optional backends: /gpu/magma /gpu/cuda/ref
Testing backends: /gpu/magma /gpu/cuda/ref
prove -v --exec 'tests/tap.sh' t536-operator t536-operator-f
t536-operator .... 
1..6
# TESTARGS: {ceed_resource}
ok 1 # SKIP - not implemented t536-operator /gpu/magma
ok 2 # SKIP - not implemented t536-operator /gpu/magma stdout
ok 3 # SKIP - not implemented t536-operator /gpu/magma stderr
not ok 4 t536-operator /gpu/cuda/ref
ok 5 t536-operator /gpu/cuda/ref stdout
not ok 6 t536-operator /gpu/cuda/ref stderr
# +/home/jeremy/Dev/libCEED/interface/ceed-elemrestriction.c:572 in CeedElemRestrictionApply(): Input vector size 48 not compatible with element restriction (48, 192)
Failed 2/6 subtests 
	(less 3 skipped subtests: 1 okay)
t536-operator-f .. 
1..6
# TESTARGS: {ceed_resource}
ok 1 # SKIP - not implemented t536-operator-f /gpu/magma
ok 2 # SKIP - not implemented t536-operator-f /gpu/magma stdout
ok 3 # SKIP - not implemented t536-operator-f /gpu/magma stderr
not ok 4 t536-operator-f /gpu/cuda/ref
ok 5 t536-operator-f /gpu/cuda/ref stdout
not ok 6 t536-operator-f /gpu/cuda/ref stderr
# +/home/jeremy/Dev/libCEED/interface/ceed-elemrestriction.c:572 in CeedElemRestrictionApply(): Input vector size 48 not compatible with element restriction (48, 192)
Failed 2/6 subtests 
	(less 3 skipped subtests: 1 okay)

Test Summary Report
-------------------
t536-operator  (Wstat: 0 Tests: 6 Failed: 2)
  Failed tests:  4, 6
t536-operator-f (Wstat: 0 Tests: 6 Failed: 2)
  Failed tests:  4, 6
Files=2, Tests=12,  5 wallclock secs ( 0.00 usr  0.01 sys +  2.59 cusr  2.06 csys =  4.66 CPU)
Result: FAIL

```

**Comment by nbeams** [comment]
Yes, but that's one of the AssembleDiagonal tests. Why is it skipped for magma and cuda-gen, but not the other cuda backends?  t530-535 are all skipped for all cuda backends, why not t536?

Actually, now I'm really confused because if I try to run t536 directly with /gpu/magma, I get
`CeedOperatorAssembleLinearDiagonal_Cuda(): Backend does not implement Operator diagonal assembly`
And the same if I try to run t535 with /gpu/cuda/ref.  What is different about t536 for /gpu/cuda/ref?

**Comment by jeremylt** [comment]
t536 has different operators. The first setup operator is causing the error.

We seem to have some crack in the logic?
```
==249559==    by 0x485BBB3: CeedErrorAbort (ceed.c:151)
==249559==    by 0x485BFD3: CeedErrorImpl (ceed.c:821)
==249559==    by 0x485D244: CeedElemRestrictionApply (ceed-elemrestriction.c:572)
==249559==    by 0x4878261: CeedOperatorApplyAdd_Cuda (ceed-cuda-operator.c:348)
==249559==    by 0x4861379: CeedOperatorApply (ceed-operator.c:933)
==249559==    by 0x109BD6: main (t536-operator.c:121)
```

**Comment by nbeams** [comment]
I think the problem could actually be in t536?  It seems it's the only triangle/nontensor test of the t530* series.  I noticed that the function `buildmats` for setting up the triangle elements has total quadrature points as `Q` (= 4) and the Strided Restrictions are set up with `Q` as the stride/element size.  But the `nqpts` variable is `nelem*Q*Q`, like in the tensor tests t534 and t535 (and the tensor tests have `Q*Q` in the Strided Restriction Create calls, too).  

When I changed `nqpts` to `nelem*Q`, all CPU backends still pass and the CUDA backends go back to the `SKIP - not implemented` output (because they're no longer failing during the setup operator apply).

**Comment by jeremylt** [comment]
I think you are right. That's as strange way for the error to come out, but its totally an error.

File: backends/cuda/ceed-cuda-operator.c (hunk 3 above)
Comment (YohannDudouit) [reply backends/cuda/ceed-cuda-operator.c]: It is classically not considered cleaner though.

**Comment by jeremylt** [comment]
@nbeams, do you want to add this fix here so we can merge without adding any failures?

File: backends/cuda/ceed-cuda-operator.c (hunk 3 above)
Comment (nbeams) [reply backends/cuda/ceed-cuda-operator.c]: @YohannDudouit do you want me to change it back?  I don't have strong feelings either way.

**Comment by nbeams** [comment]
I think 3b85b6ec21b650690c74045bd079fc6579a96e33 should take care of t536.

File: backends/cuda/ceed-cuda-operator.c (hunk 3 above)
Comment (jedbrown) [reply backends/cuda/ceed-cuda-operator.c]: No idea what "classically" means here. PETSc style uses `if (!obj)` and we use that pattern plenty in libCEED.  Whatever you choose, let's not bikeshed this.

Comment (nbeams) [reply backends/cuda/ceed-cuda-operator.c]: Let's leave it as it is.  I think the version @jeremylt suggested fits in better with the rest of libCEED.

Comment (valeriabarra) [reply backends/cuda/ceed-cuda-operator.c]: FWIW, I'm one of those who think `if (impl->evecs[i] == NULL)` is easier to read, but if the other way is preferred for consistency with the rest of the library, it makes sense to be consistent. Just saying.. Not so important

**Comment by v-dobrev** [comment]
> Thanks, @v-dobrev, that did fix the linking issue. But then I get errors like
> 
> `In function ‘int f_build_mass(void*, CeedInt, const CeedScalar* const*, CeedScalar* const*)’:`
> `examples/mfem/bp1.h:33:1: error: for, while or do statement expected before ‘case’`
> ` case 11:`
> ` ^ ~~`

I don't really understand how this error is possible (the `switch` statement is right before the `case` line printed in the error message). It looks like you found a way to fix (or work around) this issue. Or, is this still a problem?

**Comment by nbeams** [comment]
@v-dobrev I am also perplexed by that error.  I never really figured it out, but I did start a fresh configuration and build of MFEM in a new directory, which worked fine.

**Comment by nbeams** [comment]
Is it okay to squash + merge now?  I think it'd be easier to merge this first before #524, and resolve any conflicts over there.

**Pull request merged**